### PR TITLE
PCHR-2162: Open leave request popup using an unique URL

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this PR, the following feature has been developed
1. Specific Leave Request's can be opened using a unique URL

## Screencast

**Staff**
2162s

**Manager**
2162m

### URL
**Staff:** 
`/my-leave#/my-leave/report?leave-request-id=<leave-request-id>`
**Manager/Admin**:
 `/manager-leave#/manager-leave/requests?leave-request-id=<leave-request-id>`

### Approach
1. The logic to open the leave request popup has been moved from `leave-request-popup.directive.js` to newly created `leave-popup.service.js`.

2. `leave-request-popup.directive.js` has been removed, and instead `leave-popup.service.js` is called directly from all the places where we need to open the popup.

3. `LeavePopup` is a newly created service which can be used to open the leave request popup

**openModal()**
This function can be used when the leave request object is already present

**openModalByID()**
This function is used to open the leave request popup for a given ID. it checks whether the given id is present or if the logged in user has permissions to see the leave request, if not it shows an error via notification service.

4. `find` and `isManagedBy` method has been added to the `leave-request-api.js`. 
`roleOf` method has been added to `leave-request.instance.js`.

`isManagedBy` and `roleOf` has been reverted from a previous commit
https://github.com/civicrm/civihr/commit/d4173e96314d2967800614062a20c9ae855f63fc#diff-2580b71d3d2a79bd0a3154318a126cb7

5. Opening of the leave popup using an unique URL has not been implemented for the existing code.

## Comments

1. All the instances of the `uk.co.compucorp.civicrm.hrleaveandabsences` folder has been moved from `models/instances` to `instances` folder as per styleguide.
2. Angular Styleguide changes has been implemented for the touched files.
---

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/admin-dashboard/app.js [hunk 1]
```diff
@@ -26,11 +30,14 @@ define([
     'ui.router',
     'ui.select',
     'textAngular',
+    'common.angularDate',
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/admin-dashboard/app.js]: why was this added?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/app.js [hunk 2]
```diff
@@ -3,11 +3,11 @@
 define([
   'common/angular',
   'common/angularBootstrap',
-  'common/text-angular',
   'common/directives/loading',
   'common/modules/dialog',
   'common/services/angular-date/date-format',
   'common/services/check-permissions',
+  'common/text-angular',
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/app.js]: this is a vendor, it should be on top where it was before

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/admin-dashboard/app.js [hunk 3]
```diff
@@ -3,10 +3,10 @@
 define([
   'common/angular',
   'common/angularBootstrap',
-  'common/text-angular',
   'common/modules/dialog',
+  'common/text-angular',
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/admin-dashboard/app.js]: This is a vendor, it should be on top as it was before

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js [hunk 4]
```diff
@@ -183,6 +183,33 @@ define([
         },
 
         /**
+         * Get leave request for the given id
+         *
+         * @param {object} id - leave request id
+         *
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: not sure why you are adding this empty space between @param and @return, it shouldn't be here

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js [hunk 5]
```diff
@@ -183,6 +183,33 @@ define([
...
+         * @return {Promise} resolves with {Object}
+         */
+        find: function (id) {
+          $log.debug('LeaveRequestAPI.find');
+          var deferred = $q.defer();
+          var params = { id: id };
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: no point in having this, just do
```js
this.sendGet('LeaveRequest', 'get', { id: id })
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js [hunk 6]
```diff
@@ -183,6 +183,33 @@ define([
...
         /**
+         * Get leave request for the given id
+         *
+         * @param {object} id - leave request id
+         *
+         * @return {Promise} resolves with {Object}
+         */
+        find: function (id) {
+          $log.debug('LeaveRequestAPI.find');
+          var deferred = $q.defer();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: why are we using deferred at all here? can't we just do
```js
find: function (id) {
  $log.debug('LeaveRequestAPI.find');

  return this.sendGET('LeaveRequest', 'get', { id: id })
    .then(function (response) {
      if (response.values.length === 0) {
        throw new Error('LeaveRequest not found with this ID');
      }

      return response.values[0];
    };
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js [hunk 7]
```diff
@@ -183,6 +183,33 @@ define([
...
+          this.sendGET('LeaveRequest', 'get', params)
+            .then(function (response) {
+              if (response.values.length > 0) {
+                deferred.resolve(response.values[0]);
+              } else {
+                deferred.reject('LeaveRequest not found with this ID');
+              }
+            })
+            .catch(function (errorMsg) {
+              deferred.reject(errorMsg);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: isn't any error on the api level supposed to be caught by the parent `api` service?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js [hunk 8]
```diff
@@ -221,6 +248,28 @@ define([
...
 
         /**
+         * Calls the isManagedBy backend API.
+         *
+         * @param {String} leaveRequestID - ID of leave request
+         * @param {String} contactID - ID of contact
+         *
+         * @return {Promise} resolves with an {Boolean}
+         */
+        isManagedBy: function (leaveRequestID, contactID) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: I think this is still readable without having to declare a new variable
```js
isManagedBy: function (leaveRequestID, contactID) {
  $log.debug('LeaveRequestAPI.isManagedBy');

  return this.sendPOST('LeaveRequest', 'isManagedBy', {
    leave_request_id: leaveRequestID,
    contact_id: contactID
  })
  .then(function (response) {
    return response.values;
  });
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js [hunk 9]
```diff
@@ -102,12 +102,22 @@ define([
 
     vm.allowedActions = [];
 
+    vm.action = action;
+    vm.openLeavePopup = openLeavePopup;
+
+    init();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: use a IIFEE, as per the style guide

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js [hunk 10]
```diff
@@ -178,6 +181,20 @@ define([
...
+     * Opens the leave request popup
+     *
+     * @param {Object} event
+     * @param {Object} leaveRequest
+     * @param {String} leaveType
+     * @param {String} selectedContactId
+     * @param {Boolean} isSelfRecord
+     */
+    function openLeavePopup (event, leaveRequest, leaveType, selectedContactId, isSelfRecord) {
+      event.stopPropagation();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: why?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js [hunk 11]
```diff
@@ -178,6 +181,20 @@ define([
...
+      LeavePopup.openModal(leaveRequest, leaveType, selectedContactId, isSelfRecord);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: you could do this, to avoid repetition (just an idea)
```js
LeavePopup.openModel.apply(LeavePopup, arguments);
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/instances/leave-request.instance.js [hunk 12]
```diff
@@ -294,6 +294,35 @@ define([
...
+         *
+         * @param {Object} contactId
+         *
+         * @return {Promise} resolves with an {String} - manager/none
+         */
+        roleOf: function (contactId) {
+          var deferred = $q.defer();
+
+          if (this.contact_id === contactId) {
+            deferred.resolve('owner');
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/instances/leave-request.instance.js]: same comment as before, why are we using deferred?
```js
/**
 * Check the role of a given contact in relationship to the leave request.
 *
 * @param {Object} contactId
 * @return {Promise} resolves with an {String} - owner/manager/admin/none
 */
roleOf: function (contactId) {
  return (this.contact_id === contactId)
    ? $q.resolve('owner')
    : checkPermissions(sharedSettings.permissions.admin.administer)
      .then(function (isAdmin) {
        return isAdmin
          ? 'admin'
          : LeaveRequestAPI.isManagedBy(this.id, contactId)
            .then(function (isManager) {
              return isManager ? 'manager' : 'none';
            });
        }
      }.bind(this));
},
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/instances/leave-request.instance.js [hunk 13]
```diff
@@ -294,6 +294,35 @@ define([
         },
 
         /**
+         * Check the role of a given contact in relationship to the leave request.
+         *
+         * @param {Object} contactId
+         *
+         * @return {Promise} resolves with an {String} - manager/none
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/instances/leave-request.instance.js]: incorrect

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+      openModalByID: openModalByID
+    };
+
+    /**
+     * Checks if the current logged in contact can see the leave request
+     *
+     * @return {Boolean}
+     */
+    function checkPermissionBeforeOpeningPopup (leaveRequest) {
+      var deferred = $q.defer();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]:
```js
/**
 * Checks if the current logged in contact can see the leave request
 *
 * @param {LeaveRequestInstance} leaveRequest
 * @return {Promise} resolves to {Boolean}
 */
function checkPermissionBeforeOpeningPopup (leaveRequest) {
  return Session.get()
    .then(function (sessionData) {
      return leaveRequest.roleOf(sessionData.contactId);
    })
    .then(function (role) {
      return role !== 'none';
    });
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+     * Gets leave type.
+     * If leaveTypeParam exits then its a new request, else if request
+     * object exists then its edit request call
+     *
+     * @param {String} leaveTypeParam
+     * @param {Object} request leave request for edit calls
+     *
+     * @return {String} leave type
+     */
+    function getLeaveType (leaveTypeParam, request) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]:
```js
function getLeaveType (leaveTypeParam, request) {
  return request ? request.request_type : (leaveTypeParam || null);
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+      return deferred.promise;
+    }
+
+    /**
+     * Gets leave type.
+     * If leaveTypeParam exits then its a new request, else if request
+     * object exists then its edit request call
+     *
+     * @param {String} leaveTypeParam
+     * @param {Object} request leave request for edit calls
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: not `Object` but `LeaveRequestInstance`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+      return deferred.promise;
+    }
+
+    /**
+     * Gets leave type.
+     * If leaveTypeParam exits then its a new request, else if request
+     * object exists then its edit request call
+     *
+     * @param {String} leaveTypeParam
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: explain what this param is in the documentation

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+      }
+    }
+
+    /**
+     * Open leave request popup for the given leave request
+     *
+     * @param {Object} leaveRequest
+     * @param {String} leaveType
+     * @param {String} selectedContactId
+     * @param {Boolean} isSelfRecord
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: these two params *definitely* need some explanation. Would you be able to understand what they are for at a quick glance?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+     */
+    function openModal (leaveRequest, leaveType, selectedContactId, isSelfRecord) {
+      var controller = _.capitalize(getLeaveType(leaveType, leaveRequest)) + 'RequestCtrl';
+
+      $modal.open({
+        appendTo: $rootElement.children().eq(0),
+        templateUrl: sharedSettings.sharedPathTpl + 'directives/leave-request-popup/leave-request-popup.html',
+        // animation: scope.animationsEnabled,
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: commented code. Either you decomment it or you delete it

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    /**
+     * Open leave request popup for a given ID
+     *
+     * @param {String} leaveRequestID
+     * @return {Promise}
+     */
+    function openModalByID (leaveRequestID) {
+      // check if the leave request exist
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: unnecessary comment

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+        .then(function (leaveRequest) {
+          return checkPermissionBeforeOpeningPopup(leaveRequest)
+            .then(function (hasPermission) {
+              hasPermission
+                ? openModal(leaveRequest, leaveRequest.request_type, leaveRequest.contact_id, $rootScope.section === 'my-leave')
+                : notification.alert('Error', 'You dont have permission to see this leave request');
+            });
+        })
+        .catch(function (errorMsg) {
+          notification.alert('Error', errorMsg);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: what's the logic here?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+     */
+    function openModalByID (leaveRequestID) {
+      // check if the leave request exist
+      return LeaveRequest.find(leaveRequestID)
+        .then(function (leaveRequest) {
+          return checkPermissionBeforeOpeningPopup(leaveRequest)
+            .then(function (hasPermission) {
+              hasPermission
+                ? openModal(leaveRequest, leaveRequest.request_type, leaveRequest.contact_id, $rootScope.section === 'my-leave')
+                : notification.alert('Error', 'You dont have permission to see this leave request');
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: wait, don't we have a `notification.error` method?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+     * @return {Promise}
+     */
+    function openModalByID (leaveRequestID) {
+      // check if the leave request exist
+      return LeaveRequest.find(leaveRequestID)
+        .then(function (leaveRequest) {
+          return checkPermissionBeforeOpeningPopup(leaveRequest)
+            .then(function (hasPermission) {
+              hasPermission
+                ? openModal(leaveRequest, leaveRequest.request_type, leaveRequest.contact_id, $rootScope.section === 'my-leave')
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: not sure if a ternary operation is the best thing here, you're not really assigning or returning anything? a standard `if/else` would be preferrable

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/mocks/apis/leave-request-api-mock.js [hunk 24]
```diff
@@ -115,6 +115,16 @@ define([
           return $q(function (resolve, reject) {
             resolve(mockData.deleteAttachment().values);
           });
+        },
+        find: function (id) {
+          return $q(function (resolve, reject) {
+            resolve(mockData.singleDataSuccess());
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/mocks/apis/leave-request-api-mock.js]: use `$q.resolve`, don't copy from old code.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-request-actions.component.spec.js [hunk 25]
```diff
@@ -500,6 +501,28 @@ define([
...
+        event = jasmine.createSpyObj('event', ['stopPropagation']);
+        spyOn(LeavePopup, 'openModal');
+        controller.openLeavePopup(event, leaveRequest, leaveType, selectedContactId, isSelfRecord);
+      });
+
+      it('opens the leave request popup', function () {
+        expect(LeavePopup.openModal).toHaveBeenCalledWith(leaveRequest, leaveType, selectedContactId, isSelfRecord);
+      });
+
+      it('stops the vent from propagating', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-request-actions.component.spec.js]: typo

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/services/leave-popup.service.spec.js [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+      });
+
+      describe('when called with a Leave request', function () {
+        var requestType = 'SomeValue';
+
+        beforeEach(function () {
+          LeavePopup.openModal({ request_type: requestType });
+        });
+
+        it('initializes the leave popup with capitalizing the request type', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/services/leave-popup.service.spec.js]: this description doesn't have much meaning. what you are testing here is that the correct sub controller is initialized

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/services/leave-popup.service.spec.js [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+        $rootScope.$digest();
+      });
+
+      describe('when Leave Request is not found', function () {
+        beforeEach(function () {
+          LeaveRequest.find.and.returnValue($q.reject(errorMessage));
+          promise = LeavePopup.openModalByID('101');
+        });
+
+        it('initializes the leave popup with LeaveRequestController', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/services/leave-popup.service.spec.js]: wrong description, check what you are writing

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/admin-dashboard/app.js (hunk 1 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/admin-dashboard/app.js]: It is not required, removed.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/app.js (hunk 2 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/app.js]: It is not required, removed.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js (hunk 6 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: We need to reject the promise so that the chaining is done properly. Just to make sure I tried with this code, and it threw an error in the console and the chain broke.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js (hunk 7 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: Yes, but as we are returning `deferred.promise` and not the promise returned by api call, we need to reject it from here.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js (hunk 10 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: 1. It was copied from HTML
2. It is needed because the `leave-request-actions.component` sits inside `leave-request-dropdown-actions.component`. And both have on click listeners. So when the child is clicked the parent's click event should not be called.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js (hunk 11 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: In this case we cant do because of `event` param. Doing in other places.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js (hunk 19 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: Oops sorry, removed

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js (hunk 22 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: Good catch, though both the icons are same, probably that is the reason I did not notice it.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js (hunk 21 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: If the Leave request ID is not found, it shows an alert(changed to error now)

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/services/leave-popup.service.spec.js (hunk 27 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/services/leave-popup.service.spec.js]: Aplogoies

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js (hunk 10 above)
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: ok at least explain what it does in the function documentation, it's not self explanatory at all

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js (hunk 11 above)
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: why is there an `event` parameter at all?

Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: Explained in the documentation

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js (hunk 6 above)
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: instead of `throw new Error()` just return `$q.reject()` then

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js (hunk 7 above)
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: can you explain to me this in more detail? i'm still not sure i understand this

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js [hunk 28]
```diff
@@ -191,22 +191,15 @@ define([
...
+              if (response.values.length === 0) {
+                return $q.reject('LeaveRequest not found with this ID');
               }
-            })
-            .catch(function (errorMsg) {
-              deferred.reject(errorMsg);
-            });
 
-          return deferred.promise;
+              return response.values[0];
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/leave-request-api.js]: wrong indentation
